cpp/include/utilities/cython.hpp

```diff
@@ -207,6 +207,26 @@ struct random_walk_ret_t {
   std::unique_ptr<rmm::device_buffer> d_sizes_;
 };
 
+// aggregate for random_walks() COO return type
+// to be exposed to cython:
+//
+struct random_walk_coo_t {
```

Comment:
Yes, maybe reusing `cy_multi_edgelists_t` might work.

Two different structures are used here (`random_walk_coo_t`, `random_walk_ret_t`) because the RW user may want both representations (COO and coalesced paths, respectively). We don't know yet, which will gain popularity. My suspicion is that the last would. Or, maybe both would, in different contexts. One was signed off for C++ (coalesced), the other for Python (COO). But COO done in Python is expensive. Prohibitive for large graphs.

Reply:
My preference would be to collaborate with @afender to see if `cy_multi_edgelist_t` can use the more generic `num_offsets_` member instead, but that's up to the both of you. It seems worth it to be able to propagate the use of a common type IMO.

I see the potential benefit of having both RW return types (COO and coalesced paths), but my preference would be to have `call_random_walks` return the one we intend to use in python now, and remove the need to expose the converter to cython completely. If we want to switch, then we change `call_random_walks` to return the other type. Even with both types and the converter exposed, a code change is still required for the python user to use a different return type, so it doesn't really save us anything to have both.  Again, that's my preference so feel free to resolve this if we want to keep this as-is.

Reply:
I'd like to point out that `std::tuple<...>` was designed especially for this purpose: a mechanism to aggregate entities in an unnamed but flexible way. 

Unfortunately, `tuple` cannot be exposed to Cython (otherwise we wouldn't need many of these wrappers). The reason being, Cython doesn't currently support variadic templates. 

For that reason alone, a proliferation of these wrappers is inevitable. _Until_ I get to refactor all these and use the type-erasure proposed in the cascaded dispatch design (the so called _dispatch layer_; hopefully soon, but not in the next release). The gain obtained from reusing a different existent wrapper structure is overshadowed by added confusion; i.e., loss in readability and maintainability, IMHO.